# APRIORI ALGORİTMASI KULLANARAK BİRLİKTELİK KURAL ANALİZİ 

X: 1.ürün Y: 2.ürün N: Toplam Alışveriş

**- Destek (Support) : Ürünlerin Birlikte Geçme Olasılığı**

Support(X, Y) = Freq(X,Y)/N

(X ve Y farklı ürünlerinin birlikte geçme frekansı / Toplam Alışveriş)

Support değerini hesaplama amacımız eşik değeri belirlemek. Veri setine bakıldığında bir çok ürün beraber görülebilir eşik değerinin altında kalan ürünlerde eleme yapacağız.

**- Güven (Confidence) : X’i Alanların Y’yi Alma Olasılığı**

Confidence(X, Y) = Freq(X,Y)/Freq(X)

(X ve Y farklı ürünlerinin birlikte geçme frekansı / X’in gözlenme frekansı)

Analiz yapıldığında ister support değerine göre ister confidence değerine göre eşik değeri belirlenebilir. Eşik değerini gerçek hayat verilerinde çok az alabiliriz.

**- Lift = Support(X,Y)/(Support(X)*Support(Y))**

X ürünü alanların Y ürünü satın alması şu kadar kat artıyor yorumu vardır.

# Data Understanding

In [ ]:
#ROOT_DIR = "/content/drive/MyDrive/CASGEM-Egitim/Egitim-Part1/Day7-DecisionTree/notebooks"
ROOT_DIR = "https://raw.githubusercontent.com/yapay-ogrenme/casgem-eu-project-training-on-data-mining/main/PART1/Day8-UnsupervisedLearning/notebooks/"

DATASET_PATH = ROOT_DIR + "/datasets/"

In [ ]:
# .csv formatındaki veri setini okuma işlemi
# MILK,BREAD,BISCUIT olan sutün başlığını names=['products'] ile products'a çevirdik.

import pandas as pd
df = pd.read_csv(DATASET_PATH + "datasets_344_727_GroceryStoreDataSet.csv", names=['products'], header = None)
df.head()

products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

**Yapılacaklar: Veri kümesinin shape'ini kontrol ediniz.**

In [ ]:
#Yapılacaklar: Dataframe'in shape'ini kontrol ediniz.
# 20 gözlem ve 1 sütun var.
df.shape

(20, 1)

**Yapılacaklar: Veri kümesinin sütun isimlerini kontrol ediniz.**

In [ ]:
# Sutün adı ve tipi.
df.columns

Index(['products'], dtype='object')

**Yapılacaklar: Veri kümesinin her satırdaki değerlerini yazdırınız.**

In [ ]:
# Her satırda gözlenen gözlemler
df.values

array([['MILK,BREAD,BISCUIT'],
       ['BREAD,MILK,BISCUIT,CORNFLAKES'],
       ['BREAD,TEA,BOURNVITA'],
       ['JAM,MAGGI,BREAD,MILK'],
       ['MAGGI,TEA,BISCUIT'],
       ['BREAD,TEA,BOURNVITA'],
       ['MAGGI,TEA,CORNFLAKES'],
       ['MAGGI,BREAD,TEA,BISCUIT'],
       ['JAM,MAGGI,BREAD,TEA'],
       ['BREAD,MILK'],
       ['COFFEE,COCK,BISCUIT,CORNFLAKES'],
       ['COFFEE,COCK,BISCUIT,CORNFLAKES'],
       ['COFFEE,SUGER,BOURNVITA'],
       ['BREAD,COFFEE,COCK'],
       ['BREAD,SUGER,BISCUIT'],
       ['COFFEE,SUGER,CORNFLAKES'],
       ['BREAD,SUGER,BOURNVITA'],
       ['BREAD,COFFEE,SUGER'],
       ['BREAD,COFFEE,SUGER'],
       ['TEA,MILK,COFFEE,CORNFLAKES']], dtype=object)

**Yapılacaklar: Her satırdaki gözlem değerlerini virgül ile birleştirerek bir dizi içerisine tanımlayınız.**

Not: Bir satırda, birden fazla eleman(ürün) olabilir. Birlikte olanları bozmadan sadece satırları ayrı bir dizi elemanı olarak eklemelisiniz.

Birleştirme sonucu çıktı aşağıdaki gibi olmalıdır.

```
[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]
```

In [ ]:
# Her satırda tek bir gözlem birimi varmış gibiydi bunu virgül ile tek tek ayırma işlemi yaptık.
data = list(df["products"].apply(lambda x:x.split(',')))
data 

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

# Data Preprocessing

In [ ]:
from mlxtend.preprocessing import TransactionEncoder

**Yapılacaklar: One-Hot encodinge benzer şekilde TransactionEncoder() ile fit() ve transform() yaparak ürün isimlerinin sütun adlarına yazıldığı ve her bir satırın sadece içerdiği ürünlerde True diğerlerinde False yazdığı yeni bir DataFrame'e dönüştürünüz.**

**Link:** http://rasbt.github.io/mlxtend/user_guide/preprocessing/TransactionEncoder/

**Çıktı:**

| BISCUIT | BOURNVITA | BREAD | COCK | COFFEE | CORNFLAKES | JAM | MAGGI | MILK | SUGER | TEA |
|----|----|----|----|----|----|----|----|----|----|----|
|True|False|True|False|False|False|False|False|True|False|False|
|True|False|True|False|False|True|False|False|True|False|False|
.
.
.


 	 	 	 	 	 	 	 	 	


In [ ]:
#Veri kümesini istenilen True-False array'ine çevirdik.

te = TransactionEncoder()
te_data = te.fit(data).transform(data)
df = pd.DataFrame(te_data, columns=te.columns_)
df

BISCUIT  BOURNVITA  BREAD   COCK  COFFEE  CORNFLAKES    JAM  MAGGI   MILK  \
0      True      False   True  False   False       False  False  False   True   
1      True      False   True  False   False        True  False  False   True   
2     False       True   True  False   False       False  False  False  False   
3     False      False   True  False   False       False   True   True   True   
4      True      False  False  False   False       False  False   True  False   
5     False       True   True  False   False       False  False  False  False   
6     False      False  False  False   False        True  False   True  False   
7      True      False   True  False   False       False  False   True  False   
8     False      False   True  False   False       False   True   True  False   
9     False      False   True  False   False       False  False  False   True   
10     True      False  False   True    True        True  False  False  False   
11     True      False  False   True    True        True  False  False  False   
12    False       True  False  False    True       False  False  False  False   
13    False      False   True   True    True       False  False  False  False   
14     True      False   True  False   False       False  False  False  False   
15    False      False  False  False    True        True  False  False  False   
16    False       True   True  False   False       False  False  False  False   
17    False      False   True  False    True       False  False  False  False   
18    False      False   True  False    True       False  False  False  False   
19    False      False  False  False    True        True  False  False   True   

    SUGER    TEA  
0   False  False  
1   False  False  
2   False   True  
3   False  False  
4   False   True  
5   False   True  
6   False   True  
7   False   True  
8   False   True  
9   False  False  
10  False  False  
11  False  False  
12   True  False  
13  False  False  
14   True  False  
15   True  False  
16   True  False  
17   True  False  
18   True  False  
19  False   True

# Data Association Rules

In [ ]:
from mlxtend.frequent_patterns import apriori

**Yapılacaklar: apriori() fonksiyonunu kullanarak veri kümesindeki ürünlerine support değerleri ile birlikte itemsets'i hesaplattırıp yeni bir DataFrame olarak yazdırınız.**

**Link:** http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/

In [ ]:
# Apriori Fonksiyonunu deneyelim.

# df = True-False array ya da One Hot Encoding ile dönüştürülmüş dataframe
# min_support = Tüm kombinasyonların support değerini istemiyorum bu yüzden belirli bir eşik değerini getir.
# use_colnames = True, sutün isimlerini göster.

df1 = apriori(df, min_support=0.02, use_colnames=True)
df1

support                             itemsets
0      0.35                            (BISCUIT)
1      0.20                          (BOURNVITA)
2      0.65                              (BREAD)
3      0.15                               (COCK)
4      0.40                             (COFFEE)
..      ...                                  ...
78     0.05         (BISCUIT, TEA, BREAD, MAGGI)
79     0.10  (BISCUIT, COCK, CORNFLAKES, COFFEE)
80     0.05            (MAGGI, JAM, MILK, BREAD)
81     0.05             (MAGGI, TEA, BREAD, JAM)
82     0.05      (TEA, MILK, CORNFLAKES, COFFEE)

[83 rows x 2 columns]

```
BISCUIT tüm alışverişlerin % 35'inde, 
BREAD tüm alışverişlerin %65'inde veya 
TEA, MAGGI, BREAD, BISCUIT tüm satışların % 5'inde beraber yorumları yapılır.
```

**Yapılacaklar: apriori hesabı yapılan ve yeni tanımlanan dataframe'i support değerlerine göre büyükten küçüğe sıralayınızç **

In [ ]:
#Alışverişlerde en çok alınan ürünleri yorumlayabilmek için veriyi büyükten küçüğe sıralayabiliriz.

df1.sort_values(by="support", ascending=False)

support                         itemsets
2      0.65                          (BREAD)
4      0.40                         (COFFEE)
0      0.35                        (BISCUIT)
10     0.35                            (TEA)
5      0.30                     (CORNFLAKES)
..      ...                              ...
55     0.05      (BISCUIT, MILK, CORNFLAKES)
57     0.05        (SUGER, BOURNVITA, BREAD)
17     0.05                 (BISCUIT, SUGER)
37     0.05              (MAGGI, CORNFLAKES)
82     0.05  (TEA, MILK, CORNFLAKES, COFFEE)

[83 rows x 2 columns]

In [ ]:
from mlxtend.frequent_patterns import association_rules

**Yapılacaklar: Kural oluşturmak için metric olarak "confidence" i alarak minimum eşik değeri 0.6 olacak şekilde association_rules() fonksiyonunu kullanınız.**

Link: http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/

In [ ]:
# Aprioride "support" hesabı yapabiliriz, "confidence ve diğerleri" için association rules kullanıyoruz.

association_rules(df1, metric = "confidence", min_threshold = 0.6)

antecedents           consequents  antecedent support  \
0                       (COCK)             (BISCUIT)                0.15   
1                  (BOURNVITA)               (BREAD)                0.20   
2                        (JAM)               (BREAD)                0.10   
3                      (MAGGI)               (BREAD)                0.25   
4                       (MILK)               (BREAD)                0.25   
..                         ...                   ...                 ...   
81   (TEA, CORNFLAKES, COFFEE)                (MILK)                0.05   
82  (MILK, CORNFLAKES, COFFEE)                 (TEA)                0.05   
83                 (TEA, MILK)  (CORNFLAKES, COFFEE)                0.05   
84               (TEA, COFFEE)    (MILK, CORNFLAKES)                0.05   
85              (MILK, COFFEE)     (TEA, CORNFLAKES)                0.05   

    consequent support  support  confidence       lift  leverage  conviction  
0                 0.35     0.10    0.666667   1.904762    0.0475       1.950  
1                 0.65     0.15    0.750000   1.153846    0.0200       1.400  
2                 0.65     0.10    1.000000   1.538462    0.0350         inf  
3                 0.65     0.15    0.600000   0.923077   -0.0125       0.875  
4                 0.65     0.20    0.800000   1.230769    0.0375       1.750  
..                 ...      ...         ...        ...       ...         ...  
81                0.25     0.05    1.000000   4.000000    0.0375         inf  
82                0.35     0.05    1.000000   2.857143    0.0325         inf  
83                0.20     0.05    1.000000   5.000000    0.0400         inf  
84                0.10     0.05    1.000000  10.000000    0.0450         inf  
85                0.10     0.05    1.000000  10.000000    0.0450         inf  

[86 rows x 9 columns]

```
antecedent support: Birincinin tek başına görülme olasılığı,
consequent support: İkincinin tek başına görülme olasılığı,
support: İkisinin birlikte görülme olasılığı,
confidence: İlki satıldığında ikinci ürünün satılma olasılığı,
lift: İlki satıldığında ikinci ürünün satılma olasılığı şu kadar kat arttı yorumu
```

## Selecting and Filtering Results 

**Yapılacaklar: association_rules() sonucu oluşturulan kuralları "confidence" değeri 0.6'dan büyük eşit ve "support" değeri 0.2'den büyük eşit olan kuralları filtreleyiniz.**

In [ ]:
#association_rules() sonucu oluşturulan kuralları "confidence" değeri 0.6'dan büyük eşit ve 
# "support" değeri 0.2'den büyük eşit olan kuralları filtreleyiniz.

rules = association_rules(df1, metric = "confidence", min_threshold = 0.6)

rules[ (rules['confidence'] >= 0.6) & (rules['support'] >= 0.2) ]

antecedents consequents  antecedent support  consequent support  support  \
4         (MILK)     (BREAD)                0.25                0.65      0.2   
5        (SUGER)     (BREAD)                0.30                0.65      0.2   
8   (CORNFLAKES)    (COFFEE)                0.30                0.40      0.2   
9        (SUGER)    (COFFEE)                0.30                0.40      0.2   
11       (MAGGI)       (TEA)                0.25                0.35      0.2   

    confidence      lift  leverage  conviction  
4     0.800000  1.230769    0.0375        1.75  
5     0.666667  1.025641    0.0050        1.05  
8     0.666667  1.666667    0.0800        1.80  
9     0.666667  1.666667    0.0800        1.80  
11    0.800000  2.285714    0.1125        3.25

# Reporting

Analiz;
- Alışverişlerde MAGGI'ın tek başına görülme olasığı %25, TEA'in tek başına görülme olasılığı %35.
- 100 alışverişin 20'sinde mutlaka MAGGI ve TEA beraber satın alınıyor.
- MAGGI satıldığında TEA satılma olasılığı 0.800.. yani %80.
- MAGGI satılan satışlarda TEA satılma olasılığı 2.28 kat artmaktadır.


Aksiyon Fikri: 
- MAGGI alan biri %80 gibi yüksek bir ihtimalle TEA almaktadır ve TEA satışını 2.28 artırmaktadır. Bu iki ürün birbirinden uzak yerlerde konumlandırılarak müşterinin market içi dolaşması sağlanabilir, bu süreçte müşteri diğer ürünlere göz atabilir ve yahut satın alabilir. 
